### ISSUE TITLE
Permissions hierarchy doubts

### ISSUE THREAD

**Issue by maxiride** [drakkan/sftpgo#841]
Hello Drakkan,
I'm fiddling around with the permission settings, and I'm not understanding how I could achieve the following:

- A user is not allowed to upload files\create directories in its root directory
- A user is allowed to only upload files in all the subdirectories
- A user is allowed anything in a given subdirectory

Put in words, the goal is to prepare a set of directories (and subdirectories) for a user and prevent him to rename\create new (sub)directories, the user must be allowed to only upload files in any given path.
In only one subdirectory (directly child of the root path), the user can do anything.

Test configuration example I've tried, if I remove the upload permission on the root path the user can't upload anywhere (except in "libera"

```json
[...]
"permissions": {
                "/": [
                    "list",
                    "download",
                    "upload",
                    "overwrite",
                    "rename_files",
                    "delete_files"
                ],
                "/libera": [
                    "*"
                ]
            },
```

From my understanding, to achieve this I should implement permission for all the possible directories in the root one by one, but maybe I'm misunderstanding the permissions configuration.

**Comment by drakkan** [comment]
Hi,

permissions are inherited, if you remove a permission from the `/` folder you to have explictly set it on subfolders. I'm not sure how to achieve what you describe (maybe an explicit default), I'll check for this feature in other similar software

**Comment by drakkan** [comment]
I'm not sure, I have read the user guide of two software similar to SFTPGo and neither of them seems to support what you ask.

If I allow to set an explicit default instead of using parent folder permissions as the default if no explicit match is found, I have to allow it on a per directory basis. This will increase the complexity of managing permissions.

If your example the root path does not have the `create_dirs` permission so new directories cannot be created, are you serving an existing, non-empty, filesystem directory and you don't know the existing directories in advance? Can additionally directory be created, in the root path, outside SFTPGo?

**Comment by maxiride** [comment]
> are you serving an existing, non-empty, filesystem directory 

Exactly, we are migrating our FTP Server to SFTPGo and basically mapped the existing directory tree to SFTP Go (Docker based setup with Postgres data provider). Thanks to the migration tooling, I've transferred the user accounts and tweaked the transferred account's root path to match the "perceived" path inside the container from the SFTPGo point of view. So far so good and everything works as expected, we will soon do the switch.

> and you don't know the existing directories in advance? 

I do know the file system directory tree in advance, as per above. However, the directory tree is arbitrarily complex, and I would have to write down rules at least for all the first level subdirectories under `/` and exploit the inheritance to their children. As of now, indeed, given that what I want can't directly be achieved, I think I'll just write down a couple lines of code to bootstrap the permissions either through the APIs or by simply exporting/rewriting the backup file/ reimporting it.

> Can additionally directory be created, in the root path, outside SFTPGo?

Except for a "superuser" account, whose account root is the SFTPGo data root with access to all the user accounts, directories would be created outside SFTPGo directly in the underlying host, in this case a local file system.


---
To give more context, we heavily rely on software automation to perform tasks on files deposited on the FTP. Some users use automated software to do it, guaranteeing reproducibility, but others manually use FTP Clients (like FileZilla) to upload files and very often delete/rename folders on which **our** infrastructure depends.
The goal is to "lock" the directory tree to prevent such issues, except for a dedicated "exchange" directory for spot use cases and transfers.

**Comment by drakkan** [comment]
Hi,

using the group feature (available only in the development version for now), you can set/update the same permissions for different accounts just updating the group.

In your case, a pre-login hook might be useful

https://github.com/drakkan/sftpgo/blob/main/docs/dynamic-user-mod.md

you could list the root directory and set the permissions for top level directories in missing for the user trying to login

**Comment by maxiride** [comment]
> you could list the root directory and set the permissions for top level directories if missing for the user trying to login

This will do, thanks for the hint! I must have missed the feature while exploring the application.

**Comment by drakkan** [comment]
This [change](https://github.com/drakkan/sftpgo/commit/2017cb60e90c25aadf0a5eb9a04d2de8a38ac56b) could simplify your setup

**Comment by maxiride** [comment]
Will look in the updates! Thanks for keeping me posted 😀

**Comment by drakkan** [comment]
I think it will be in v2.5, but you can already test it using the development version
